test/e2e/storage/csi_mock_volume.go

```diff
@@ -730,6 +743,202 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
+
+		createVolume := "CreateVolume"
+		deleteVolume := "DeleteVolume"
+		publishVolume := "NodePublishVolume"
+		unpublishVolume := "NodeUnpublishVolume"
+		stageVolume := "NodeStageVolume"
+		unstageVolume := "NodeUnstageVolume"
+
+		// These calls are assumed to occur in this order for
+		// each test run.
+		deterministicCalls := []string{
+			createVolume,
+			stageVolume,
+			publishVolume,
+			unpublishVolume,
+			unstageVolume,
+			deleteVolume,
+		}
+
+		for _, t := range tests {
+			test := t
+			ginkgo.It(test.name, func() {
+				var err error
+				params := testParameters{
+					lateBinding:    test.lateBinding,
+					enableTopology: test.topology,
+
+					// Not strictly necessary, but without it, the external-attacher takes two minutes to detach
+					// the volume?! Looks like a bug.
+					disableAttach:  true,
+					registerDriver: true,
+				}
+
+				if test.resourceExhausted {
+					params.javascriptHooks = map[string]string{
+						"globals": `counter=0; console.log("globals loaded", OK, INVALIDARGUMENT)`,
+						// Every second call returns RESOURCEEXHAUSTED, starting with the first one.
+						"createVolumeStart": `console.log("Counter:", ++counter); if (counter % 2) { RESOURCEEXHAUSTED; } else { OK; }`,
+					}
+				}
+
+				init(params)
+				defer cleanup()
+
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				pvcWatch, err := f.ClientSet.CoreV1().PersistentVolumeClaims(f.Namespace.Name).Watch(ctx, metav1.ListOptions{})
+				framework.ExpectNoError(err, "create PVC watch")
+				defer pvcWatch.Stop()
+
+				sc, claim, pod := createPod(false)
+				gomega.Expect(pod).NotTo(gomega.BeNil(), "while creating pod")
+				bindingMode := storagev1.VolumeBindingImmediate
+				if test.lateBinding {
+					bindingMode = storagev1.VolumeBindingWaitForFirstConsumer
+				}
+				framework.ExpectEqual(*sc.VolumeBindingMode, bindingMode, "volume binding mode")
+
+				err = e2epod.WaitForPodNameRunningInNamespace(m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "failed to start pod")
+				err = e2epod.DeletePodWithWait(m.cs, pod)
+				framework.ExpectNoError(err, "failed to delete pod")
+				err = m.cs.CoreV1().PersistentVolumeClaims(claim.Namespace).Delete(context.TODO(), claim.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "failed to delete claim")
+
+				normal := []csiCall{}
+				for _, method := range deterministicCalls {
+					normal = append(normal, csiCall{expectedMethod: method})
+				}
+				expected := normal
+				// When simulating limited capacity,
+				// we expect exactly two CreateVolume
+				// calls because the first one should
+				// have failed.
+				if test.resourceExhausted {
+					expected = []csiCall{
+						{expectedMethod: createVolume, expectedError: codes.ResourceExhausted},
+					}
+					expected = append(expected, normal...)
+				}
+
+				var calls []mockCSICall
+				err = wait.Poll(time.Second, csiPodRunningTimeout, func() (done bool, err error) {
```

Comment:
By this time, test pvc/pod has been created, run and deleted. So all calls must be in mock driver logs. Isn't 5 minute timeout too much to get the logs?

Reply:
I've replaced this with a per-test timeout. It would be even nicer if that deadline could also be passed into the other helper functions (like `WaitForPodNameRunningInNamespace`) but those have their own builtin timeouts.

@jsafrane: please have another look